documentation/specs/static-graph.md

```diff
@@ -143,49 +139,100 @@ This is a proposal for what the public API may look like:
         }
     }
 
-## Inferring which targets to run and cache for a project within the graph
+## Inferring which targets to run for a project within the graph
 
-One property of the static build graph is that a project (project, globalproperties, toolsversion) is "built" / visited only once when the project graph is executed.
+One property of the static build graph is that a project (project, globalproperties, toolsversion) is "built" / visited only once when the project graph is executed. This is in conflict with current P2P patterns (project to project) where a project may be built multiple times during a build with different targets and global properties (examples in [p2p protocol](https://github.com/Microsoft/msbuild/blob/master/documentation/ProjectReference-Protocol.md)).
 
-This is in conflict with current P2P patterns (project to project) where a project may be built multiple times during a build with different targets and global properties (examples in [p2p protocol](https://github.com/Microsoft/msbuild/blob/master/documentation/ProjectReference-Protocol.md)). Thus we need a mechanism through which the project files self describe all the ways they can be built in. We need a public interface of msbuild targets w.r.t. the build graph. We call this interface the **exported targets**.
+Additionaly, projects are built bottom up, whereas the classical msbuild scheduler builds projects top down. In the top down traversal, the parent project chooses which targets to call on the child projects. But in the bottom up traversal, we don't know the entry targets for a child because we visit the children before the parents.
 
-Executing projects with export targets within the graph means changing the behavior of the MSBuild task to always pull the target results from cache rather than actually evaluating and executing the list targets. This has several benefits such as each project having a concrete lifetime within the build, may reduce re-evaluation cost, and opens up maybe caching and distribution scenarios.
+We need to represent project to project calling patterns in such a way that a graph build can infer the entry targets for a project. These protocols apply to an entire graph. They can be regular or irregular. **Regular** operations consist of recursively calling the same target on all children (e.g. build, clean, rebuild). **Irregular** operations consist of calling different targets on different nodes in the graph (e.g. publish calls publish on the root node but build on the rest). As a complication, today nothing prevents projects to turn regular operations into irregular operations (e.g. a project file reimplements Clean by calling Clean2 on children).
```

Comment:
Is the regular/irregular distinction required? I find it confusing and it doesn't seem to be relevant, since I think you could make the whole thing work based on single-level target knowledge.